paper_framework_sources/p1897_initial_executor_algorithms_for_cpp23.md

```diff
@@ -419,66 +453,76 @@ int r = sync_wait(just(3));
-int r = sync_wait(just(3) | via(t.scheduler()));
+int r = sync_wait(just(3) | on(my_scheduler{}) | transform([](int v){return v+1;}));
 // r==3
 ```
+*- end example]*
+
 
 ### Wording
-The name `execution::via` denotes a customization point object.
-The expression `execution::via(S, Sch)` for some subexpressions `S`, `Sch` is expression-equivalent to:
+The name `execution::on` denotes a customization point object.
+For some subexpressions `s` and `sch`, let `S` be a type such that `decltype((s))` is `S` and `Sch` be a type such that `decltype((sch))` is `Sch`
+The expression `execution::on(s, sch)` is expression-equivalent to:
 
- * `S.via(Sch)` if that expression is valid.
- * Otherwise, `via(S, Sch)` if that expression is valid with overload resolution performed in a context that includes the declaration
+ * `s.on(sch)` if that expression is valid, if `S` satisfies `sender`
+ * Otherwise, `on(s, sch)` if that expression is valid, and if `S` satisfies `sender` and if `Sch` satisfies `scheduler`, with overload resolution performed in a context that includes the declaration
  ```
-         template<class S, class Sch>
-           void via(S, Sch) = delete;
+      void on() = delete;
  ```
+   and that does not include a declaration of `execution::on`.
 
- * Otherwise constructs a receiver `r` such that when `set_value`, `set_error` or `set_done` is called on `r` the value(s) or error(s) are packaged, and a receiver `r2` constructed such that when `execution::set_value(r2)` is called, the stored value or error is transmitted and `r2` is submitted to `Sch`. If `set_error` or `set_done` is called on `r2` the error or cancellation is propagated and the packaged values ignored.
- * The returned sender's value types match those of `sender1`.
- * The returned sender's execution context is that of `scheduler1`.
+ * Otherwise:
+
+   * Constructs an operation such that when `connect` is called with a `receiver` `output_receiver`:
+     * Constructs a receiver, `r` such that when `set_value`, `set_error` or `set_done` is called on `r`, the parameter is wrapped in a receiver `r2`.
+      * `r2` is passed to `execution::connect(execution::schedule(sch), std::move(r2))`.
+       * `execution::start` is passed the resulting `operation_state`.
+       * When `set_value` is called on `r2`, the stored value is forwarded to `output_receiver`.
+       * When `set_error` or `set_done` is called on `r2` the parameters propagate to `output_receiver`.
+     * Passes `r` to `execution::connect(s, r)` resulting in an operation state `ros`.
+   * When `execution::start` is called on the resulting `operation_state`, call `execution::start(ros)`.
+     * Calls `execution::start` on the resulting `operation_state`.
+   * If `output_receiver` satisfies `scheduler_provider` then `execution::get_scheduler(r2)` returns the result of `execution::get_scheduler(output_receiver)`.
+   * `r` satisfies `scheduler_provider` and `execution::get_scheduler(r2)` returns sch.
```

Comment:
I think this is an 1898 question really so I will not change 1897 for it.

However, it'd be good to understand this. Is your goal here to connect the scheduler chain early, so that in a chain like:

on(s1) | alg1 | alg2 | on(s2) | alg3 | on(s3)

a receiver that returns s2 from get_scheduler will not be submitted into s1 until alg2 completes.

I think customising on is roughly what my example code in https://github.com/facebookexperimental/libunifex/pull/124 does - because it does the reservation at connect time. Your extension would hook schedulers together at connect time.